src/main.ts

```diff
@@ -0,0 +1,258 @@
+  }
+
+  if (yargsFailure) {
+    const { message } = yargsFailure;
+    outputLogger.logToStderr(
+      'ERROR: %s\n\n%s',
+      message,
+      await parser.getHelp(),
+    );
+    return null;
+  }
+
+  return {
+    ...options,
+    // This is the name that Yargs gives to the arguments that aren't options.
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    _: options._.map(String),
+  };
+}
+
+/**
+ * Runs all of the rules on the given projects.
+ *
+ * This is a bit complicated because linting may fail for a particular project.
+ * In that case, we don't want to crash the whole tool, but fail gracefully.
+ *
+ * @param args - The arguments to this function.
+ * @param args.projectReferences - References to the projects (either the name
+ * of a MetaMask repository, such as "utils", or the path to a local Git
+ * repository).
+ * @param args.template - The repository to which the project should be
+ * compared.
+ * @param args.workingDirectoryPath - The directory where this tool was run.
+ * @param args.cachedRepositoriesDirectoryPath - The directory where MetaMask
+ * repositories will be (or have been) cloned.
+ * @param args.outputLogger - Writable streams for output messages.
+ */
+async function lintProjects({
+  projectReferences,
+  template,
+  workingDirectoryPath,
+  cachedRepositoriesDirectoryPath,
+  outputLogger,
+}: {
+  projectReferences: string[];
+  template: MetaMaskRepository;
+  workingDirectoryPath: string;
+  cachedRepositoriesDirectoryPath: string;
+  outputLogger: OutputLogger;
+}) {
+  const projectLintResultPromiseOutcomes = await Promise.allSettled(
+    projectReferences.map(async (projectReference) => {
+      return await lintProject({
+        projectReference,
+        template,
+        rules,
+        workingDirectoryPath,
+        cachedRepositoriesDirectoryPath,
+        outputLogger,
+      });
+    }),
+  );
+
+  const fulfilledProjectLintResultPromiseOutcomes =
+    projectLintResultPromiseOutcomes.filter(isPromiseFulfilledResult);
+  const rejectedProjectLintResultPromiseOutcomes =
+    projectLintResultPromiseOutcomes.filter(isPromiseRejectedResult);
+
+  outputLogger.logToStdout('');
+  fulfilledProjectLintResultPromiseOutcomes
+    .sort((a, b) => {
+      return a.value.projectName.localeCompare(b.value.projectName);
+    })
+    .forEach((fulfilledProjectLintResultPromiseOutcome, index) => {
+      reportProjectLintResult({
+        projectLintResult: fulfilledProjectLintResultPromiseOutcome.value,
+        outputLogger,
+      });
+      if (index < fulfilledProjectLintResultPromiseOutcomes.length - 1) {
+        outputLogger.logToStdout('');
+      }
+    });
+  outputLogger.logToStdout('');
+
+  rejectedProjectLintResultPromiseOutcomes.forEach(
+    (rejectedProjectLintResultPromiseOutcome) => {
+      outputLogger.logToStderr(
+        'stack' in rejectedProjectLintResultPromiseOutcome.reason
+          ? rejectedProjectLintResultPromiseOutcome.reason.stack
```

Comment:
This seems.... a little strange to me. Why are we logging just the stack here, rather than the full error? `console.error` at least seems to format errors reasonable well (including the stack), and it automatically supports error causes too which is a bonus

Reply:
That's fair! I've fixed the OutputLogger to handle error objects (well, non-strings in general) better: [d034718](https://github.com/MetaMask/module-lint/pull/33/commits/d034718c7f27b8a6ed1e24e48c6b89a855edbd29)